### PR TITLE
Base Styles: Import colors into variables

### PR DESCRIPTION
**Pull request by ockham**
## Description

`@import "./colors";` from within `_variables.scss`, since the latter [depends](https://github.com/WordPress/gutenberg/blob/3687e8b992b6462044eead19944cf0c6ea41a1e3/packages/base-styles/_variables.scss#L36-L40) on `$dark-gray-900` which is [defined](https://github.com/WordPress/gutenberg/blob/bfd1117846c7802d62dbba046c27968d961781e7/packages/base-styles/_colors.scss#L7) in the former.

Found while working on https://github.com/Automattic/wp-calypso/pull/38445.

## How has this been tested?

Try importing _just_ `@wordpress/base-styles/variables` in your app, using the currently released package. You'll get a build error saying something like

```
    Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
    SassError: Undefined variable: "$dark-gray-900".
```

Now try with the changes that this PR makes. Your style files should build successfully :+1: 

## Types of changes

Bug fix.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by simison** [comment]
Hmm right, I suppose elsewhere we've always been importing all the files at once so that all variables always exist. I.e. as highlighted in docs:

https://github.com/WordPress/gutenberg/tree/bfd1117846c7802d62dbba046c27968d961781e7/packages/base-styles#scss-utilities-and-variables

Things obviously won't work when importing just one file directly like in your PR.

The build/test step during CI runs won't catch these either — should it?

**Comment by ockham** [comment]
CircleCI's Lint job is failing, complaining that it

```
 Expected double quotes
```

Which I find a bit surprising? Is Gutenberg using double quotes in SASS imports?

**Comment by ockham** [comment]
> Hmm right, I suppose elsewhere we've always been importing all the files at once so that all variables always exist. I.e. as highlighted in docs:
> 
> https://github.com/WordPress/gutenberg/tree/bfd1117846c7802d62dbba046c27968d961781e7/packages/base-styles#scss-utilities-and-variables
> 
> Things obviously won't work when importing just one file directly like in your PR.
> 
> The build/test step during CI runs won't catch these either — should it?

I think that'd make sense :+1:

**Comment by simison** [comment]
> lint error [...] Which I find a bit surprising? Is Gutenberg using double quotes in SASS imports?

Yeah — Gutenberg prolly https://www.npmjs.com/package/stylelint-config-wordpress which is double quotes indeed, if I remember correct.

**Comment by ockham** [comment]
> > lint error [...] Which I find a bit surprising? Is Gutenberg using double quotes in SASS imports?
> 
> Yeah — Gutenberg prolly https://www.npmjs.com/package/stylelint-config-wordpress which is double quotes indeed, if I remember correct.

Pushed a fix :+1:

**Comment by simison** [comment]
> > The build/test step during CI runs won't catch these either — should it?
>
>I think that'd make sense 👍

I personally think we should render each file separately as a test run and ensure;
- ensure they don't fatal 
- they don't produce any visible CSS

More convo about testing in original PR https://github.com/WordPress/gutenberg/pull/17883 — see commits for details. I don't think that needs to be done in this PR tho.

File: packages/base-styles/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## master
+
+- Import `colors` into `variables` since the latter depends on the former.
```
Comment (gziolo) [inline packages/base-styles/CHANGELOG.md]:
```suggestion
### Bug Fix

- Import `colors` into `variables` since the latter depends on the former.
```

**Review by gziolo — APPROVED** [review]
It would be nice to include a section title in the changelog to make it easier to read. Thanks for the fix 💯

**Comment by hypest** [comment]
It seems that the native mobile side is broken by this change, with this error:
```
Error: ./colors could not be resolved in /home/circleci/project/gutenberg/packages/edit-post/src/components/layout,/home/circleci/project,gutenberg/packages/edit-post/src/components/layout,/home/circleci/project/src,/home/circleci/project/gutenberg/assets/stylesheets,/home/circleci/project/gutenberg/packages/base-styles
```

My first impression is that the SASS transformer we use in the RN app pipeline doesn't understand partial SASS files, and is unable to resolve the (existing) `./_colors.scss`.

**Comment by simison** [comment]
@hypest is the solution to import `./_colors.scss` (i.e. include the file extension) in Gutenberg or fix something at your side?

**Comment by hypest** [comment]
👋 @simison!

On the Gutenberg-web side, including _the underscore_ is fixing it for native. The native mobile SASS transformer apparently doesn't support partial SASS files so, it doesn't automatically try to resolve the underscore-less file to the underscored one.

If handled on the native mobile side alone, a solution seems to be to introduce a new file `colors.native.scss`, and in it just `@import "./_colors"`. It's a bit hacky but, can be a workaround.

I wonder though, perhaps it's an obsolete impression but, I thought that the `colors.scss` files were automatically imported for all SASS files so, one didn't need to explicitly import them 🤔. That's what we had implemented in native mobile too, automatically import `_colors.scss` and others.

**Comment by ockham** [comment]
> I wonder though, perhaps it's an obsolete impression but, I thought that the `colors.scss` files were automatically imported for all SASS files so, one didn't need to explicitly import them . That's what we had implemented in native mobile too, automatically import `_colors.scss` and others.

The point of this PR was to enable 3rd party projects to import individual files from `@wordpress/base-styles` (specifically, `_variables.scss`) by making the -- otherwise implicit -- styling dependency chain explicit. Without the colors import, just importing `_variables.scss` wouldn't work, since the 3rd party project wouldn't know where to find the definition for `$dark-gray-900`, which is used in `_variables.scss`.

**Comment by ockham** [comment]
@hypest Are you sure about the partials support thing? I thought that the underscore/non-underscore aliasing was built into SASS at a pretty low level (but I might be wrong).

**Comment by ockham** [comment]
> It seems that the native mobile side is broken by this change, with this error:
> 
> ```
> Error: ./colors could not be resolved in /home/circleci/project/gutenberg/packages/edit-post/src/components/layout,/home/circleci/project,gutenberg/packages/edit-post/src/components/layout,/home/circleci/project/src,/home/circleci/project/gutenberg/assets/stylesheets,/home/circleci/project/gutenberg/packages/base-styles
> ```
> 
> My first impression is that the SASS transformer we use in the RN app pipeline doesn't understand partial SASS files, and is unable to resolve the (existing) `./_colors.scss`.

@hypest Do you have some more context on this (e.g. failing CI job link)? :slightly_smiling_face: 
It's kinda weird that it's not resolving the file correctly...

**Comment by ockham** [comment]
Ah I think I found the relevant file https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js

**Comment by hypest** [comment]
> @hypest Are you sure about the partials support thing? I thought that the underscore/non-underscore aliasing was built into SASS at a pretty low level (but I might be wrong).

I'm not too sure, I've only assuming that the SASS transformer we use in the RN mobile port (this library: https://github.com/kristerkari/react-native-sass-transformer) doesn't support it because if I just add the underscore the import is resolved OK.


> @hypest Do you have some more context on this (e.g. failing CI job link)? 🙂
> It's kinda weird that it's not resolving the file correctly...

The only CI indication that this fails is on the gutenberg-mobile repo for now. See this https://app.circleci.com/jobs/github/wordpress-mobile/gutenberg-mobile/20081 and search for "colors could not be resolved".

**Comment by ockham** [comment]
Looks like it's coming from gutenberg-mobile itself: https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js#L139

Can we maybe just make that thing smarter? https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js#L133

**Comment by ockham** [comment]
> Looks like it's coming from gutenberg-mobile itself: https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js#L139

(I guess that's our local fork of https://github.com/kristerkari/react-native-sass-transformer)

**Comment by hypest** [comment]
> Can we maybe just make that thing smarter? https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js#L133

Perhaps. It feels that'd be equivalent to trying to add support for partials 😬. Maybe we can live with just trying out the underscored filename.

> (I guess that's our local fork of https://github.com/kristerkari/react-native-sass-transformer)

Indeed.

**Comment by ockham** [comment]
> > Can we maybe just make that thing smarter? https://github.com/wordpress-mobile/gutenberg-mobile/blob/06b924d832e1bdd5bcbb2961f8c5e2051d772269/sass-transformer.js#L133
> 
> Perhaps. It feels that'd be equivalent to trying to add support for partials :grimacing: 

Reading https://sass-lang.com/guide, I got the impression that there isn't that much to partials, other than a nomenclature to signal that they shouldn't be turned into CSS on their own (but used as imports elsewhere), and the leading underscore alias resolution :man_shrugging: 

> Maybe we can live with just trying out the underscored filename.

Yeah, that might work as well.

> > (I guess that's our local fork of https://github.com/kristerkari/react-native-sass-transformer)
> 
> Indeed.

:+1:

**Comment by ockham** [comment]
Turns out upstream already has support for partials: https://github.com/kristerkari/react-native-sass-transformer/blob/52884dd59582856fa17e2b2e8ca9efc37d412387/index.js#L41

**Pull request merged**